### PR TITLE
[Server] Multicluster: Feed back context persistence

### LINKED ISSUE #5873: [UX] Multi-Cluster: Improve cluster connection showback
#### Current Behavior
Upon uploading kubeconfig(s) in Meshery UI, Meshery Server will attempt to connect to contexts within the kubeconfig. Users are shown a snackbar upon successful addition of a cluster, but without any context-specific details.

#### Desired Behavior
Ideally, users get one snackbar per context added or ignored. Each notification should include the context name and address.

#### Acceptance Tests
1. "Upload" confirmation snackbar is changed to Informational (blue-colored) message.
1. One success (green) notification is displayed per new context added.
1. One warning (yellow) notification is displayed per context ignored (not added).
1. User sees confirmation of the number of contexts added and number of contexts ignored.

_More Context:_

> Yes, so in that case we fall back to not flattening. This is included

For users, this will be in almost all cases. We need two things:

1. Further instructions in the modal on how to prepare your kubeconfig.
1. Additional snackbar notifications containing details of the context(s) that was just added and/or details of context(s) that were skipped over.

_Originally posted by @leecalcote in https://github.com/meshery/meshery/issues/5867#issuecomment-1192817843_

Screen Shot 2022-07-22 at 4 53 41 PM

---

#### Contributor [Guides](https://docs.meshery.io/project/contributing) and Resources

- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🖥 [Contributing to Meshery UI](https://docs.meshery.io/project/contributing/contributing-ui)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by gr455**
**Description**
Meshery server now responds with the summary of persistence of k8s contexts. The response contains `InsertedContexts`, `UpdatedContexts` and `ErroredContexts` arrays containing k8s contexts that were inserted, updated and errored while inserting respectively.

To do:
- [x] Make corresponding snackbars for Inserted Contexts (green), Updated Contexts (yellow) and Errored Contexts (red) on the UI 

This PR fixes #5873

Screenshot from 2022-08-02 20-36-04

**Notes for Reviewers**
Also review the corresponding PR in meshery cloud

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by gr455** [comment]
@leecalcote Does this look okay? I can then move forward with making the snackbars

File: models/providers.go [hunk 1]
```diff
@@ -122,6 +122,12 @@ type Capability struct {
 	Endpoint string  `json:"endpoint,omitempty"`
 }
 
+// K8sContextResponse - struct of response sent by provider when requested to persist k8s config
+type K8sContextPersistResponse struct {
+	K8sContext
+	Inserted bool
```
Comment (Revolyssup) [inline models/providers.go:128]: add json tag

File: handlers/k8sconfig_handler.go [hunk 2]
```diff
@@ -25,6 +25,13 @@ import (
 	"github.com/spf13/viper"
 )
 
+// SaveK8sContextResponse - struct used as (json marshalled) response to requests for saving k8s contexts
+type SaveK8sContextResponse struct {
```
Comment (Revolyssup) [inline handlers/k8sconfig_handler.go:29]: Add json tags

**Review by Revolyssup — CHANGES_REQUESTED** [review]

**Comment by gr455** [comment]
@AakashRaj20 do you think you can handle the UI part of this, or would you want to try a good first issue to start?

What we want are individual snackbars for each kubernetes context. The kubernetes contexts are returned by the server on uploding the kubeconfig. In this PR, the response has the format
```json
{
"inserted_contexts": [],
"updated_contexts": [],
"errored_contexts": [],
}
```
Each context in those arrays has two feilds - `name` and `server`
we want to display both the name and the server in the snackbars for each context.
The color of the snackbars would be 
- Green for inserted_contexts (along with success message)
- Yellow for updated_contexts (along with a message saying the context already existed)
- Red for errored_contexts (along with a failure message)

**Comment by leecalcote** [comment]
> @AakashRaj20 do you think you can handle the UI part of this, or would you want to try a good first issue to start?
> 
> What we want are individual snackbars for each kubernetes context. The kubernetes contexts are returned by the server as a response on uploding the kubeconfig. In this PR, the response has the format
> 
> ```json
> {
> "inserted_contexts": [],
> "updated_contexts": [],
> "errored_contexts": [],
> }
> ```
> 
> Each context in those arrays has two fields - `name` and `server` we want to display both the name and the server in the snackbars for each context. The color of the snackbars would be
> 
> * Green for inserted_contexts (along with success message)
> * Yellow for updated_contexts (along with a message saying the context already existed)
> * Red for errored_contexts (along with a failure message)

@YashKamboj, perhaps, you can knock this out?

**Comment by AakashRaj20** [comment]
> > @AakashRaj20 do you think you can handle the UI part of this, or would you want to try a good first issue to start?
> > 
> > What we want are individual snackbars for each kubernetes context. The kubernetes contexts are returned by the server as a response on uploding the kubeconfig. In this PR, the response has the format
> > 
> > ```json
> > {
> > "inserted_contexts": [],
> > "updated_contexts": [],
> > "errored_contexts": [],
> > }
> > ```
> > 
> > Each context in those arrays has two fields - `name` and `server` we want to display both the name and the server in the snackbars for each context. The color of the snackbars would be
> > 
> > * Green for inserted_contexts (along with success message)
> > * Yellow for updated_contexts (along with a message saying the context already existed)
> > * Red for errored_contexts (along with a failure message)
> 
> @YashKamboj, perhaps, you can knock this out?

Hey @leecalcote so sorry for the late reply I had my external labs for college if its still okay I can try the UI part in this issue or it can be a new issue.

**Comment by leecalcote** [comment]
@AakashRaj20, I understand. Have you contributed to Meshery UI previously? If not, have you compiled Meshery Server and Meshery UI?

**Comment by YashKamboj** [comment]
@leecalcote Ok, i will tak a look into it

**Comment by gr455** [comment]
@YashKamboj Any updates on this?

**Comment by leecalcote** [comment]
@YashKamboj we’ll want to bump this to the top of your priority list.

**Comment by gr455** [comment]
Merge the PR in meshery-cloud before merging this one

File: ui/components/MeshConfigComponent.js [hunk 3]
```diff
@@ -257,6 +257,49 @@ function MesherySettingsNew({ classes, enqueueSnackbar, closeSnackbar, updatePro
     });
   }
 
+  const handleConfigSnackbars = ctxs => {
+    updateProgress({ showProgress : false });
+
+    for (let ctx of ctxs.inserted_contexts) {
+      const msg = `Cluster ${ctx.name} at ${ctx.server} added successfully`
```
Comment (leecalcote) [inline ui/components/MeshConfigComponent.js:264]:
```suggestion
      const msg = `Cluster ${ctx.name} at ${ctx.server} connected`
```

File: ui/components/MeshConfigComponent.js [hunk 4]
```diff
@@ -257,6 +257,49 @@ function MesherySettingsNew({ classes, enqueueSnackbar, closeSnackbar, updatePro
...
+          <IconButton key="close" aria-label="Close" color="inherit" onClick={() => closeSnackbar(key)}>
+            <CloseIcon />
+          </IconButton>
+        ),
+        autoHideDuration : 2000,
+      });
+    }
+
+    for (var ctx of ctxs.updated_contexts) {
+      const msg = `Cluster ${ctx.name} at ${ctx.server} already exists`
```
Comment (leecalcote) [inline ui/components/MeshConfigComponent.js:277]: Please consider this as an info, not a warning.

Comment (gr455) [reply ui/components/MeshConfigComponent.js:277]: Alright :+1:

**Comment by gr455** [comment]
@leecalcote This is also ready to be merged

**Comment by leecalcote** [comment]
Refreshing…

**Comment by gr455** [comment]
@leecalcote Ready to merge?

**Comment by leecalcote** [comment]
I missed this. SGTM

**Pull request merged**